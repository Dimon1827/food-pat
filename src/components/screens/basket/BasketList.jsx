import { useDispatch, useSelector } from 'react-redux';
import {
  clearAll,
  selectProducts,
  deleteSelectedProducts,
} from '../../../redux/slice/productsSlice';
import BasketItem from './BasketItem';
import Button from '../../ui/button/Button';

function BasketList() {
  const products = useSelector(selectProducts);
  const dispatch = useDispatch();

  const handleClearAllClick = () => {
    dispatch(clearAll());
  };

  const handleClearSelectedClick = () => {
    dispatch(deleteSelectedProducts());
  };

  return (
    <div>
      {products.length > 0 ? (
        <>
          <ul className="basket__list">
            {products.map((product, index) => (
              <BasketItem key={index} {...product} />
            ))}
          </ul>
          <div className="basket__wrapper-button">
            <Button className="basket__button" onClick={handleClearAllClick}>
              Очистить всё
            </Button>
            <Button
              className="basket__button"
              onClick={handleClearSelectedClick}
            >
              Удалить выбранные товары
            </Button>
          </div>
        </>
      ) : (
        <span className="basket__empty">Корзина пуста</span>
      )}
    </div>
  );
}

export default BasketList;
