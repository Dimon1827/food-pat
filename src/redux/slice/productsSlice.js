import { createSlice } from '@reduxjs/toolkit';

const initialState = [];

const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    addProduct: (state, action) => {
      state.push(action.payload);
    },
    deleteProduct: (state, action) => {
      return state.filter(({ id }) => action.payload !== id);
    },
    clearAll: () => {
      return [];
    },
  },
});

export const { addProduct, deleteProduct, clearAll } = productsSlice.actions;

export const selectProducts = (state) => state.products;

export default productsSlice.reducer;
