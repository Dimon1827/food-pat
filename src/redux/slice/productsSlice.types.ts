import { IProduct } from '@/types';

export interface IProducts
  extends Omit<IProduct, 'ingredients' | 'description'> {
  isChecked: boolean;
  quantity: number;
}

export interface IReduxProduct {
  products: IProducts[];
  totalPrice: number;
  totalQuantity: number;
}
