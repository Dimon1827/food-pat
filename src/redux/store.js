import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import productsReducer from './slice/productsSlice';

const persistConfig = {
  key: 'root',
  version: 2,
  storage,
};

const reducer = combineReducers({
  products: productsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistor = persistStore(store);

export default store;
