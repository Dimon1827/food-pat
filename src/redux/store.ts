import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import productsReducer from './slice/productsSlice';

const persistConfig = {
  key: 'root',
  version: 2,
  storage,
};

const reducer = combineReducers({
  products: productsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export type RootState = ReturnType<typeof persistedReducer>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const persistor = persistStore(store);
export default store;
