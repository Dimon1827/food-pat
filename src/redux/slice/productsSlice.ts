import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IProducts, IReduxProduct } from './productsSlice.types';

const initialState: IReduxProduct = {
  products: [],
  totalPrice: 0,
  totalQuantity: 0,
};

const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    addProduct: (state, action: PayloadAction<IProducts>) => {
      state.products.push(action.payload);
    },
    deleteProduct: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        products: state.products.filter(({ id }) => action.payload !== id),
      };
    },
    toggleProduct: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        products: state.products.map((product) => {
          return product.id === action.payload
            ? { ...product, isChecked: !product.isChecked }
            : product;
        }),
      };
    },
    deleteSelectedProducts: (state) => {
      return {
        ...state,
        products: state.products.filter((product) => !product.isChecked),
      };
    },
    clearAll: () => {
      return {
        products: [],
        totalPrice: 0,
        totalQuantity: 0,
      };
    },
    incrementQuantity: (state, action: PayloadAction<string>) => {
      const product = state.products.find((item) => item.id === action.payload);
      if (product) {
        product.quantity++;
      }
    },
    decrementQuantity: (state, action: PayloadAction<string>) => {
      const product = state.products.find((item) => item.id === action.payload);
      if (product && product.quantity > 1) {
        product.quantity--;
      }
    },
    calculateAll: (state) => {
      const { total, quantity } = state.products.reduce(
        (productsTotal, product) => {
          const { price, quantity } = product;
          const productTotal = price * quantity;

          productsTotal.total += productTotal;
          productsTotal.quantity += quantity;

          return productsTotal;
        },
        { total: 0, quantity: 0 }
      );
      state.totalPrice = total;
      state.totalQuantity = quantity;
    },
  },
});

export const {
  addProduct,
  deleteProduct,
  clearAll,
  toggleProduct,
  deleteSelectedProducts,
  incrementQuantity,
  decrementQuantity,
  calculateAll,
} = productsSlice.actions;

export default productsSlice.reducer;
