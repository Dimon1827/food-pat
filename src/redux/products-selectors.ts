import { RootState } from "./store";

export const selectProducts = (state: RootState) => state.products.products;
export const selectTotalPrice = (state: RootState) => state.products.totalPrice;
export const selectTotalQuantity = (state: RootState) => state.products.totalQuantity;
