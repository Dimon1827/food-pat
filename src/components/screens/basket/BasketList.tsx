import { useSelector } from 'react-redux';
import {
  clearAll,
  deleteSelectedProducts,
  calculateAll,
} from '../../../redux/slice/productsSlice';
import BasketItem from './BasketItem';
import Button from '../../ui/button/Button';
import { useEffect } from 'react';
import {
  selectProducts,
  selectTotalPrice,
} from '../../../redux/products-selectors';
import { useAppDispatch } from '../../../redux/store';

function BasketList() {
  const products = useSelector(selectProducts);
  const total = useSelector(selectTotalPrice);
  const dispatch = useAppDispatch();

  const handleClearAllClick = () => {
    dispatch(clearAll());
  };

  const handleClearSelectedClick = () => {
    dispatch(deleteSelectedProducts());
  };

  useEffect(() => {
    dispatch(calculateAll());
  }, [dispatch, products]);

  return (
    <div className="basket__main-wrapper">
      {products.length > 0 ? (
        <>
          <ul className="basket__list">
            {products.map((product, index) => (
              <BasketItem key={index} {...product} />
            ))}
          </ul>
          <span className="basket__total">Общая сумма: {total} ₽ </span>
          <div className="basket__wrapper-button">
            <Button className="basket__button" onClick={handleClearAllClick}>
              Очистить всё
            </Button>
            <Button
              className="basket__button"
              onClick={handleClearSelectedClick}
            >
              Удалить выбранные товары
            </Button>
          </div>
        </>
      ) : (
        <span className="basket__empty">Корзина пуста</span>
      )}
    </div>
  );
}

export default BasketList;
