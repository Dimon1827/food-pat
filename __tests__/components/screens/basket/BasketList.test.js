import BasketList from '@/components/screens/basket/BasketList';
import * as actions from '@/redux/slice/productsSlice';
import * as store from '@/redux/store';
import { fireEvent, render, screen } from '@testing-library/react';
import * as reduxHooks from 'react-redux';
import { selectProducts, selectTotalPrice } from '@/redux/products-selectors';

jest.mock('react-redux');
const dispatch = jest.fn();
const mockedDispatch = jest.spyOn(store, 'useAppDispatch');
const mockedSelector = jest.spyOn(reduxHooks, 'useSelector');
const ArrayProducts = [
  {
    id: '80e90d1a-31d4-43dd-9eee-60a3bc2f56ca',
    name: 'Диетическая курица',
    price: 600,
    image: '/assets/images/product-image-3',
    isChecked: false,
    quantity: 1,
  },
  {
    id: '80e90d1a-31d4-43dd-9eee-60a3bc2f56hf',
    name: 'Диетическая курица',
    price: 600,
    image: '/assets/images/product-image-2',
    isChecked: false,
    quantity: 1,
  },
];

describe('BasketList', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    mockedSelector.mockImplementation((selector) => {
      if (selector === selectProducts) {
        return ArrayProducts;
      }
      if (selector === selectTotalPrice) {
        return 600;
      }
      return null;
    });
  });

  it('BasketList should render correctly', () => {
    mockedDispatch.mockReturnValue(dispatch);
    render(<BasketList />);
    expect(screen.getByTestId('basket-list')).toMatchSnapshot();
  });
  it('Button should clear all elements', () => {
    mockedDispatch.mockReturnValue(dispatch);
    render(<BasketList />);
    const mockClearAll = jest.spyOn(actions, 'clearAll');
    const ClearAllButton = screen.getByTestId('clear-all');
    fireEvent.click(ClearAllButton);
    expect(dispatch).toHaveBeenCalled();
    expect(mockClearAll).toHaveBeenCalledTimes(1);
  });
  it('Button should delete selected products', () => {
    mockedDispatch.mockReturnValue(dispatch);
    render(<BasketList />);
    const mockDeleteSelectedProducts = jest.spyOn(
      actions,
      'deleteSelectedProducts'
    );
    const deleteSelectedButton = screen.getByText('Удалить выбранные товары');
    fireEvent.click(deleteSelectedButton);
    expect(dispatch).toHaveBeenCalled();
    expect(mockDeleteSelectedProducts).toHaveBeenCalledTimes(1);
  });
  it('Should be appear text basket is empty if products is zero', () => {
    mockedDispatch.mockReturnValue(dispatch);
    mockedSelector.mockImplementation((selector) => {
      if (selector === selectProducts) {
        return [];
      }
      if (selector === selectTotalPrice) {
        return 0;
      }
      return null;
    });

    render(<BasketList />);
    const text = screen.getByText('Корзина пуста');
    expect(text).toBeInTheDocument();
  });
});
