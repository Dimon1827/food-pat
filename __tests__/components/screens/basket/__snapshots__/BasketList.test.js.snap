// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketList BasketList should render correctly 1`] = `
<div
  class="basket__main-wrapper"
  data-testid="basket-list"
>
  <ul
    class="basket__list"
  >
    <li
      class="basket__item"
    >
      <div
        class="basket__toggle "
        data-testid="toggle-button"
      />
      <img
        alt="Диетическая курица"
        class="basket__image"
        src="/assets/images/product-image-3.jpg"
      />
      <div
        class="basket__wrapper"
      >
        <h2
          class="basket__subtitle"
        >
          Диетическая курица
        </h2>
        <span
          class="basket__price"
        >
          600
           ₽
        </span>
        <svg
          class="basket__dustbin"
          data-testid="delete-button"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
          />
        </svg>
        <div
          class="basket__quantity-wrapper"
        >
          <button
            class="basket__decrement"
            data-testid="button-decrement"
          >
            <svg
              viewBox="0 0 24 24"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
          <div
            class="basket__quantity"
          >
            1
          </div>
          <button
            class="basket__increment"
            data-testid="button-increment"
          >
            <svg
              viewBox="0 0 24 24"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
        </div>
      </div>
    </li>
    <li
      class="basket__item"
    >
      <div
        class="basket__toggle "
        data-testid="toggle-button"
      />
      <img
        alt="Диетическая курица"
        class="basket__image"
        src="/assets/images/product-image-2.jpg"
      />
      <div
        class="basket__wrapper"
      >
        <h2
          class="basket__subtitle"
        >
          Диетическая курица
        </h2>
        <span
          class="basket__price"
        >
          600
           ₽
        </span>
        <svg
          class="basket__dustbin"
          data-testid="delete-button"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
          />
        </svg>
        <div
          class="basket__quantity-wrapper"
        >
          <button
            class="basket__decrement"
            data-testid="button-decrement"
          >
            <svg
              viewBox="0 0 24 24"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
          <div
            class="basket__quantity"
          >
            1
          </div>
          <button
            class="basket__increment"
            data-testid="button-increment"
          >
            <svg
              viewBox="0 0 24 24"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
        </div>
      </div>
    </li>
  </ul>
  <span
    class="basket__total"
  >
    Общая сумма: 
    600
     ₽ 
  </span>
  <div
    class="basket__wrapper-button"
  >
    <button
      class="button basket__button"
      data-testid="clear-all"
    >
      Очистить всё
    </button>
    <button
      class="button basket__button"
    >
      Удалить выбранные товары
    </button>
  </div>
</div>
`;
